### ISSUE TITLE
Dependent filter selects first value independently of the filter it depends on, if it is set to "Select first filter value as default"

### ISSUE THREAD

**Issue by e-geist** [apache/superset#22979]
Hi,

we experienced that using required filters in Dashboards that cascade/depend on each other with "first value" as default lead to invalid filter combinations upon first dashboard load.

If a filter "events" exists whose values are defined as dependent on a filter "year" and both filters are
- required
- set to "Select first value by default"

upon loading a dashboard, the first value in the filter "year" will be picked (as supposed) and the first value in the filter "events" will be picked - even if the event doesn't exist for the picked year.
The Charts are then queried with this invalid filter combination and lead to "No data" - until the user fixes the picked filters. 

#### How to reproduce the bug

1. Create a dashboard with two filters and some charts these filters are applied to
2. Make both filters required and "Select first filter value as default"
3. Make one of the filters dependent on the other
4. Make sure the first value of the dependent filter is not available with the first value of the other filter
5. Load dashboard in UI
6. First value of both filters is chosen automatically -> invalid combination of filters is chosen -> "No data" in charts

### Expected results

The dependent filters' value is chosen dependent on the other filters' value - so:
First value of first filter is selected -> second filters' values are filtered according to first filter's value -> second filters' filtered list is used to choose first value

### Actual results

Both filters' values are chosen indepedently from each other.

### Environment
- browser type and version: Firefox
- superset version: 2.0.1
- python version: Python 3.8.12
- any feature flags active: nothing too special - we use embedding dashboards, but this also occurs for normal dashboards

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.
- [x] I have asked other people in Superset Slack - it seems to be reproducible also for others (see here: https://apache-superset.slack.com/archives/C016B3LG5B4/p1675352073631509)

If you have any further questions or anything is unclear, please let me know.

**Comment by sfirke** [comment]
I also experience this and would just add: once the user makes a selection for the 2nd variable, they are correctly restricted from choosing an invalid combination.  I.e., the filter 2 choices correctly depend on filter 1.  I hope that means this would be fixable.

**Comment by fmannhardt** [comment]
I have encountered the same issue. Unfortunately, anyone using the "required value" setting and a cascading filter is forced to select a default value leading to this bug. 

In my case, I would already be helped with there would be the option to have "required value" set without forcing me to specify a default value. In this case the behavior would be simply to show no results in case the required value was not yet set by the user.

**Comment by IgDem** [comment]
I confirm this issue. It makes useless shared filters.

**Comment by HeathLee** [comment]
any update for this bug?

**Comment by rusackas** [comment]
> any update for this bug?

Nope... this is where you'd hear news if there was any. Please feel free to investigate a fix. We'd love to see a PR from anyone on the thread if it's a problem.

**Comment by surapuramakhil** [comment]
@rusackas I would like to take this issue.

**Comment by rusackas** [comment]
> @rusackas I would like to take this issue.

Sounds great! Good luck, and thank you in advance!

**Comment by anantaoutlook** [comment]
@rusackas,  I have fixed this issue. Please assign this to me to create PR

**Comment by sfirke** [comment]
@anantaoutlook feel free to create a pull request that mentions this issue.

I see @surapuramakhil volunteered to work on this. Have you made progress? I'd be very happy to have this closed by either of you, just want to avoid having two fix PRs hit at the same time.

**Comment by anantaoutlook** [comment]
@sfirke I dont have a access to create a branch here

**Comment by sfirke** [comment]
@anantaoutlook okay I assigned you. And it's okay to send a PR from a branch in your own forked repo too.

**Comment by anantaoutlook** [comment]
@sfirke, thank you for assigning this to me. I still dont see a way to create a branch. Am I missing anything?

**Comment by sfirke** [comment]
From the main repository page, click fork near the top to create your own copy of the repository. From there you can create a new branch, make your changes there, and when you are ready create a pull request to merge changes from your forked branch back to the main master branch. Does that work?

**Comment by surapuramakhil** [comment]
@sfirke @anantaoutlook is this issue fixed? I could see above 3 MR's are closed.

**Comment by sadpandajoe** [comment]
> @sfirke @anantaoutlook is this issue fixed? I could see above 3 MR's are closed.

@surapuramakhil I think it might still be an issue. Those PRs were created and closed and never merged, and it looks like there is now another PR open: https://github.com/apache/superset/pull/30122.

**Comment by vigodeltoro** [comment]
Hi
issue isn't fixed. I was testing the PRs and they will lead to new bugs..

**Comment by rusackas** [comment]
Is anyone still planning to address this? Maybe it's resolved in the latest releases, or maybe it's an edge case that not many people are facing, but this has been silent for quite a while and is at risk of being closed as stale/not-planned.

**Comment by janaperedni** [comment]
Hi! Still experiencing this issue:(((((

**Comment by Pilario** [comment]
Ive the same issue in Version: 4.1.2

Why do we even need set a default value for a required field? If no value, graphs shouldnt be rendered fe

**Comment by dipoppe** [comment]
Ive the same issue in Version: 4.1.2

Why do we even need set a default value for a required field? If no value, graphs shouldnt be rendered fe

**Comment by rusackas** [comment]
This bug continues to be reported against recent versions (4.1.2 as of June 2025). It appears to be a legitimate issue with how dependent filters initialize their default values. If anyone is interested in contributing a fix, that would be very welcome — this has been outstanding for quite a while and affects real users.

**Comment by vigodeltoro** [comment]
Hi @rusackas as I was informed from my colleague this was fixed by our commit and can be closed.

https://github.com/apache/superset/pull/34137

So in my eyes this should be already fixed ( beginning with the versions where the merge is included )

**Comment by vigodeltoro** [comment]
And it is working in our production :)

**Comment by sfirke** [comment]
Thanks @vigodeltoro for the update! I will close accordingly.
